Sources/Rendering/OpenGL/VolumeMapper/index.js

```diff
@@ -376,16 +389,82 @@ function vtkOpenGLVolumeMapper(publicAPI, model) {
       needRebuild = true;
     }
 
+    const numComp = model.scalarTexture.getComponents();
+    const iComps = actor.getProperty().getIndependentComponents();
+    let usesProportionalComponents = false;
+    const proportionalComponents = [];
+    if (iComps) {
+      // Define any proportional components
+      for (let nc = 0; nc < numComp; nc++) {
+        proportionalComponents.push(actor.getProperty().getOpacityMode(nc));
+      }
+
+      if (proportionalComponents.length > 0) {
+        usesProportionalComponents = true;
+      }
+    }
+
+    const ext = model.currentInput.getExtent();
+    const spc = model.currentInput.getSpacing();
+    const vsize = vec3.create();
+    vec3.set(
+      vsize,
+      (ext[1] - ext[0]) * spc[0],
+      (ext[3] - ext[2]) * spc[1],
+      (ext[5] - ext[4]) * spc[2]
+    );
+
+    const maxSamples =
+      vec3.length(vsize) / model.renderable.getSampleDistance();
+
+    const state = {
+      interpolationType: actor.getProperty().getInterpolationType(),
+      useLabelOutline: actor.getProperty().getUseLabelOutline(),
+      numComp,
+      usesProportionalComponents,
+      iComps,
+      maxSamples,
+      useGradientOpacity: actor.getProperty().getUseGradientOpacity(0),
+      blendMode: model.renderable.getBlendMode(),
+      averageIPScalarMode: model.renderable.getAverageIPScalarRange(),
+      proportionalComponents,
+    };
+
+    // We only need to rebuild the shader if one of these variables has changed,
+    // since they are used in the shader template replacement step.
+    if (
+      !model.previousState ||
+      model.previousState.interpolationType !== state.interpolationType ||
+      model.previousState.useLabelOutline !== state.useLabelOutline ||
+      model.previousState.numComp !== state.numComp ||
+      model.previousState.usesProportionalComponents !==
+        state.usesProportionalComponents ||
+      model.previousState.iComps !== state.iComps ||
+      model.previousState.maxSamples !== state.maxSamples ||
+      model.previousState.useGradientOpacity !== state.useGradientOpacity ||
+      model.previousState.blendMode !== state.blendMode ||
+      !arrayEquals(
+        model.previousState.averageIPScalarMode,
+        state.averageIPScalarMode
+      ) ||
+      !arrayEquals(
+        model.previousState.proportionalComponents,
+        state.proportionalComponents
+      )
+    ) {
```

Comment:
hey @jourdain @swederik Can the above block of code be simplified as something like 
```
if (model.previousState === state)
```

My knowledge of JS is rough but it seems like it could be a simple object to object comparison maybe?

Reply:
or @floryst , really anyone with more of a clue than me :-)

Reply:
Unfortunately `model.previousState === state` won't be enough as it will always be false (not same ref between calls).

But if everything is serializable we could do `JSON.stringify(model.previousState) !== JSON.stringify(state)`